repair/writer.hh

```diff
@@ -23,22 +25,31 @@ class mutation_fragment_queue {
 public:
     class impl {
     public:
-        virtual future<> push(mutation_fragment mf) = 0;
+        virtual future<> push(mutation_fragment_v2 mf) = 0;
         virtual void abort(std::exception_ptr ep) = 0;
         virtual void push_end_of_stream() = 0;
         virtual ~impl() {}
     };
 
 private:
     std::unique_ptr<impl> _impl;
+    lw_shared_ptr<std::vector<future<>>> _pending;
+    upgrading_consumer<std::function<void(mutation_fragment_v2)>> _consumer;
 
 public:
-    mutation_fragment_queue(std::unique_ptr<impl> impl)
+    mutation_fragment_queue(schema_ptr s, reader_permit permit, std::unique_ptr<impl> impl)
         : _impl(std::move(impl))
+        , _pending(make_lw_shared(std::vector<future<>>()))
+        , _consumer(*s, std::move(permit), [this](mutation_fragment_v2 mf) {
+            _pending->push_back(_impl->push(std::move(mf)));
```

Comment:
I don't think this is legal. Unless otherwise specified, if a method returns a future, you may not call it again, before the future returned from the previous invocation was waiter for.
I think `_pending` should be a `std::vector<mutation_fragment_v2>` and `push()` should take care of flushing it, waiting for each fragment as it is pushed.

Reply:
upgrading_consumer expects to receive a mutation_fragment_v2 and return void. It's job is to pass the fragment to impl. impl->push returns a future, which I can't co_await on because of the upgrading consumer signature.

Should I enforce that impl->push co_awaits internally so that the push is flushed and then returns a ready future so that I can ignore it via something like:

```
class impl {
     void push(mutation_fragment_v2 mf) {
        auto result = do_push(std::move(mf));
        assert(result.available());
     }
     virtual void do_push(mutation_fragment_v2 mf) = 0;
};
```

If that's the case, then _pending is unnecessary and I can remove it altogether. If not, please explain where I misunderstood your comment.

Reply:
You cannot return a ready future by `co_await`-ing internally. This still translates to returning an unready future if the co-awaited operation blocks.
I think you should push into an internal buffer in the consumer like I suggested, and do the possibly blocking flush into `_impl` in `push()`, where you *can* wait for each fragment's future.